flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java

```diff
@@ -0,0 +1,579 @@
+
+	/**
+	 * We disable writes to the write-ahead-log here. We can't have these in the base class
+	 * because JNI segfaults for some reason if they are.
+	 */
+	private final WriteOptions writeOptions;
+
+	/**
+	 * Creates a new {@code RocksDBMapState}.
+	 *
+	 * @param namespaceSerializer The serializer for the namespace.
+	 * @param stateDesc The state identifier for the state.
+	 */
+	public RocksDBMapState(ColumnFamilyHandle columnFamily,
+			TypeSerializer<N> namespaceSerializer,
+			MapStateDescriptor<UK, UV> stateDesc,
+			RocksDBKeyedStateBackend<K> backend) {
+
+		super(columnFamily, namespaceSerializer, stateDesc, backend);
+
+		this.userKeySerializer = stateDesc.getKeySerializer();
+		this.userValueSerializer = stateDesc.getValueSerializer();
+
+		writeOptions = new WriteOptions();
+		writeOptions.setDisableWAL(true);
+	}
+
+	// ------------------------------------------------------------------------
+	//  MapState Implementation
+	// ------------------------------------------------------------------------
+
+	@Override
+	public UV get(UK userKey) throws IOException {
+		try {
+			byte[] rawKeyBytes = serializeUserKeyWithCurrentKeyAndNamespace(userKey);
+			byte[] rawValueBytes = backend.db.get(columnFamily, rawKeyBytes);
+
+			return (rawValueBytes == null ? null : deserializeUserValue(rawValueBytes));
+		} catch (RocksDBException e) {
+			throw new RuntimeException("Error while getting data from RocksDB.", e);
+		}
+	}
+
+	@Override
+	public void put(UK userKey, UV userValue) throws IOException {
+		if (userValue == null) {
+			remove(userKey);
+		}
+
+		try {
+			byte[] rawKeyBytes = serializeUserKeyWithCurrentKeyAndNamespace(userKey);
+			byte[] rawValueBytes = serializeUserValue(userValue);
+
+			backend.db.put(columnFamily, writeOptions, rawKeyBytes, rawValueBytes);
+		} catch (RocksDBException e) {
+			throw new RuntimeException("Error while putting data into RocksDB", e);
+		}
+	}
+
+	@Override
+	public void remove(UK userKey) throws IOException {
+		try {
+			byte[] rawKeyBytes = serializeUserKeyWithCurrentKeyAndNamespace(userKey);
+
+			backend.db.remove(columnFamily, writeOptions, rawKeyBytes);
+		} catch (RocksDBException e) {
+			throw new RuntimeException("Error while removing data from RocksDB.", e);
+		}
+	}
+
+	@Override
+	public boolean contains(UK userKey) throws IOException {
+		try {
+			byte[] rawKeyBytes = serializeUserKeyWithCurrentKeyAndNamespace(userKey);
+			byte[] rawValueBytes = backend.db.get(columnFamily, rawKeyBytes);
+
+			return (rawValueBytes != null);
+		} catch (RocksDBException e) {
+			throw new RuntimeException("Error while getting data from RocksDB", e);
+		}
+	}
+
+	@Override
+	public int size() throws IOException {
+		Iterator<Map.Entry<UK, UV>> iterator = iterator();
+
+		int count = 0;
+		while (iterator.hasNext()) {
+			count++;
+			iterator.next();
+		}
+
+		return count;
+	}
+
+	@Override
+	public Iterable<UK> keys() {
+		return new Iterable<UK>() {
```

Comment:
The `Iterable` is used to iterate over the mappings under the current key. Though we can reuse a singleton object for different keys, it will make the implementation complicated.